### PR TITLE
Support k8s 1.7.x

### PR DESCRIPTION
**Pull request by samba**
Relates to samsung-cnct/k2#607

This is partially working. For me it correctly brings up a 1.7.1 cluster, and I can deploy stateless Nginx to it. Known issues described below. 

Feel free to suggests improvements if anything stands out.

### CODE REVIEW

**Comment by samba** [comment]
This revision does not currently:

- [ ] Deploy the "vanilla" Kubernetes dashboard
- [ ] Install Helm chart for either GitLab or MongoDB w/ stateful sets 

There appears to be some issue with the way storage classes (e.g. the default) are configured in the K2 cluster, so the Helm charts I've attempted will:

- Deploy a `PersistentVolumeClaim`, which gets stuck in `Pending` state, because:
- No persistent volumes can be provisioned/claimed/etc.

I've attempted all this *even after* deploying the below storage class, to no avail. 

```yaml
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: standard
  annotations:
    storageclass.kubernetes.io/is-default-class: "true"
    volume.alpha.kubernetes.io/storage-class: "default"
provisioner: kubernetes.io/aws-ebs
parameters:
  type: gp2
```

File: ansible/roles/kraken.config/files/config.yaml [hunk 1]
```diff
@@ -61,6 +61,12 @@ definitions:
       type: canal
       options:
         containers:
+          kubePolicyController:
+            version: v0.5.1
+            location: calico/kube-policy-controller
+          etcd:
+            version: v3.0.9
+            location: quay.io/coreos/etcd
```
Comment (mikeln) [inline ansible/roles/kraken.config/files/config.yaml]: This is the wrong setup.  from 1.6 on we use the non-etcd (self contained) setup.  I thought this was documented in the 1.6 template file. (which is baed on this setup: https://github.com/projectcalico/canal/blob/master/k8s-install/1.6/canal.yaml).  Earlier version used this external etcd setup.

Do not add these lines.

File: ansible/roles/kraken.config/files/config.yaml [hunk 2]
```diff
@@ -77,6 +83,34 @@ definitions:
...
+        containers:
+          kubePolicyController:
+            version: v0.6.0
+            location: calico/kube-policy-controller
+          etcd:
+            version: v3.0.9
+            location: quay.io/coreos/etcd
+          calicoCni:
+            version: v1.9.1
+            location: quay.io/calico/cni
```
Comment (mikeln) [inline ansible/roles/kraken.config/files/config.yaml]: same comment here, as above.  this should be exactly the same as the current 1.6.

File: ansible/roles/kraken.fabric/kraken.fabric.flannel/templates/v1.7/canal.yaml.part.jinja2 [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+          configMap:
+            name: canal-config
+
+---
+
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: canal
+  namespace: kube-networking
```
Comment (mikeln) [inline ansible/roles/kraken.fabric/kraken.fabric.flannel/templates/v1.7/canal.yaml.part.jinja2]: this is correct.  notice there is no policyController or etcd backing store.

File: ansible/roles/kraken.nodePool/kraken.nodePool.master/templates/v1.7/api-server-manifest.yaml.jinja2 [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    - --etcd-certfile=/etc/etcd/ssl/client.pem
+    - --etcd-keyfile=/etc/etcd/ssl/client-key.pem
+    - --allow-privileged=true
+    - --service-cluster-ip-range={{ cluster.network }}
+    - --secure-port=443
+    - --admission-control=NamespaceLifecycle,LimitRanger,ServiceAccount,ResourceQuota
+    - --tls-cert-file=/etc/kubernetes/ssl/apiserver.pem
+    - --tls-private-key-file=/etc/kubernetes/ssl/apiserver-key.pem
+    - --client-ca-file=/etc/kubernetes/ssl/ca.pem
+    - --basic-auth-file=/etc/kubernetes/basicauth.csv
```
Comment (mikeln) [inline ansible/roles/kraken.nodePool/kraken.nodePool.master/templates/v1.7/api-server-manifest.yaml.jinja2]: You need to rebase your branch.  
There are recent changes that enable/disable basicauth that you are missing here.

File: ansible/roles/kraken.rbac/templates/rbac-default-policy.yaml.jinja2 [hunk 5]
```diff
@@ -121,6 +121,18 @@ roleRef:
...
+- kind: User
+  name: {{ cluster.name }}_admin
+roleRef:
+  kind: ClusterRole
+  name: cluster-admin
+  apiGroup: ""
+---
+kind: ClusterRoleBinding
+apiVersion: rbac.authorization.k8s.io/v1beta1
+metadata:
```
Comment (mikeln) [inline ansible/roles/kraken.rbac/templates/rbac-default-policy.yaml.jinja2]: why is this needed?  This is redefining a built-in role.  
our default_basic_user:  has to be in the system:masters group.  There should not need to be a user of this type defined in the default rbac.   This comes back to my question, what is actually wrong with the default user being named `admin` :  `system:masters`?

File: ansible/roles/kraken.services/templates/rbac-dns-policy.yaml.jinja2 [hunk 6]
```diff
@@ -1,7 +1,7 @@
 ---
 # kubedns Roles
 kind: ClusterRole
-apiVersion: rbac.authorization.k8s.io/v1alpha1
+apiVersion: rbac.authorization.k8s.io/v1beta1
```
Comment (mikeln) [inline ansible/roles/kraken.services/templates/rbac-dns-policy.yaml.jinja2]: this is correct only if pre v1.6 can't get to this file.   (which I think is true, as each release has it's own subdir).

File: ansible/roles/kraken.services/templates/v1.7/dns.yaml.jinja2 [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+          operator: "Exists"
+      volumes:
+        - name: kube-dns-config
+          configMap:
+            name: {{dns_config.k8sapp}}
+            optional: true
+      containers:
+        - name: kubedns
+          image: {{dns_config.kubedns_image_v16}}
+          resources:
```
Comment (mikeln) [inline ansible/roles/kraken.services/templates/v1.7/dns.yaml.jinja2]: needs to be kubedns_image_v17

File: ansible/roles/kraken.services/templates/v1.7/dns.yaml.jinja2 [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+            protocol: TCP
+          - containerPort: 10055
+            name: metrics
+            protocol: TCP
+          volumeMounts:
+          - name: kube-dns-config
+            mountPath: /kube-dns-config
+        - name: dnsmasq
+          image: {{dns_config.dnsmasq_image_v16}}
+          livenessProbe:
```
Comment (mikeln) [inline ansible/roles/kraken.services/templates/v1.7/dns.yaml.jinja2]: needs to be dnsmasq_image_v17

File: ansible/roles/kraken.services/templates/v1.7/dns.yaml.jinja2 [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+          resources:
+            requests:
+              cpu: {{dns_config.dnsmasq_cpu_req}}
+              memory: {{dns_config.dnsmasq_mem_req}}
+          volumeMounts:
+          - name: kube-dns-config
+            mountPath: /etc/k8s/dns/dnsmasq-nanny
+        - name: sidecar 
+          image: {{dns_config.dnsmasq_metrics_image_v16}} 
+          livenessProbe:
```
Comment (mikeln) [inline ansible/roles/kraken.services/templates/v1.7/dns.yaml.jinja2]: needs to be dnsmasq_metrics_image_v17

File: ansible/roles/kraken.services/defaults/main.yaml [hunk 10]
```diff
@@ -26,6 +27,7 @@ kubedns:
   kubedns_cpu_req: 100m
   kubedns_mem_req: 70Mi
   healthz_image: "gcr.io/google_containers/exechealthz-amd64:1.2"
+  healthz_image_v17: "gcr.io/google_containers/exechealthz-amd64:1.2"
```
Comment (mikeln) [inline ansible/roles/kraken.services/defaults/main.yaml]: delete this line.  healthz is not uased as a separate image since 1.6+

File: ansible/roles/kraken.access/tasks/generate-admin-kubeconfig.yaml [hunk 11]
```diff
@@ -26,7 +26,7 @@
 - name: create admin user entry from basicauth user
   command: >
     {{ kubectl }} config set-credentials {{ cluster.name }}_admin
-      --username={{ default_basic_kube_auth.user }}
+      --username={% if cluster.kubeAuth.authz.rbac %}{{ cluster.name }}_admin{% else %}{{ default_basic_kube_auth.user }}{% endif %}
```
Comment (samba) [inline ansible/roles/kraken.access/tasks/generate-admin-kubeconfig.yaml]: With 1.7, drawing from the 1.6 configuration previous, RBAC is set up and Ansible fails even to `get nodes` during its verification phase, so it quits. 

To get past that, I discovered (by manually altering my own kubeconfig) that setting the username to the generated one  `{{ cluster.name }}_admin` would allow me to authenticate and continue the deployment. This approach has worked reproducibly ever since.

If there's an alternate approach you'd suggest, I'm open to it, but this is the only way I found that worked... so far.

File: ansible/roles/kraken.config/files/config.yaml (hunk 1 above)
Comment (samba) [inline ansible/roles/kraken.config/files/config.yaml]: Hmm, somehow these were required; it kept failing on missing the key `etcd` during deployment. I'll take another look at that and see if I can find a way to proceed without them.

File: ansible/roles/kraken.services/defaults/main.yaml [hunk 12]
```diff
@@ -15,6 +15,7 @@ kubedns:
   version: v17
   color: FFFF66
   kubedns_image: "gcr.io/google_containers/kubedns-amd64:1.8"
+  kubedns_image_v17: "gcr.io/google_containers/k8s-dns-kube-dns-amd64:1.14.4"
   kubedns_image_v16: "gcr.io/google_containers/k8s-dns-kube-dns-amd64:1.14.2"
   kubedns_image_v15: "gcr.io/google_containers/kubedns-amd64:1.9"
   kubedns_image_v14: "gcr.io/google_containers/kubedns-amd64:1.8"
```
Comment (coffeepac) [inline ansible/roles/kraken.services/defaults/main.yaml]: 1.4 support is dropped with adding 1.7

File: ansible/roles/kraken.services/defaults/main.yaml [hunk 13]
```diff
@@ -26,6 +27,7 @@ kubedns:
   kubedns_cpu_req: 100m
   kubedns_mem_req: 70Mi
   healthz_image: "gcr.io/google_containers/exechealthz-amd64:1.2"
+  healthz_image_v17: "gcr.io/google_containers/exechealthz-amd64:1.2"
   healthz_image_v15: "gcr.io/google_containers/exechealthz-amd64:1.2"
   healthz_image_v14: "gcr.io/google_containers/exechealthz-amd64:1.2"
```
Comment (coffeepac) [inline ansible/roles/kraken.services/defaults/main.yaml]: same, will not repeat comment again but all 1.4 should be removed.

File: ansible/roles/kraken.nodePool/kraken.nodePool.node/templates/v1.7/units.kubelet.part.jinja2 [hunk 14]
```diff
@@ -0,0 +1,62 @@
+- name: kubelet.service
```
Comment (coffeepac) [inline ansible/roles/kraken.nodePool/kraken.nodePool.node/templates/v1.7/units.kubelet.part.jinja2]: is this the same as the 1.6 file?   I think we talked about this briefly in person.  if so, maybe we move the 1.5 file (currently default) into a versioned directory and let this be the default.  

this applies to the networking configs as well.

**Comment by coffeepac** [comment]
a.  thanks!  
b. if you have the time, can you make the default version of the files be the common file?  ie:  if the common file is for 1.6 & 1.7, please make that the default.  this will require checking that 1.5 still works so it is some scope creep.  if you don't have time for this, please file an issue so someone else can perform this cleanup and link it here.
c. please set the default version of kubernetes to be 1.7.1 so that CI can test it.  this PR is using 1.6.7 in CI right now.  you may also need to push a temporary k2-tools image (and use it in the Jenkinsfile) while samsung-cnct/k2-tools#97 keeps moving

File: ansible/roles/kraken.config/files/config.yaml (hunk 1 above)
Comment (mikeln) [inline ansible/roles/kraken.config/files/config.yaml]: That would indicate that your setup is reading the wrong template dir...  (i.e. v1.7 subdir is missing, in which case the default (1.4) is being read).

File: ansible/roles/kraken.access/tasks/generate-admin-kubeconfig.yaml (hunk 11 above)
Comment (mikeln) [inline ansible/roles/kraken.access/tasks/generate-admin-kubeconfig.yaml]: We really need to understand why for this one.  The behavior is odd.

File: ansible/roles/kraken.config/files/config.yaml (hunk 1 above)
Comment (samba) [inline ansible/roles/kraken.config/files/config.yaml]: Thanks for the tip. I'm sure that was true when I made this change as I hadn't yet created the 1.7 directories. Will reevaluate.

**Comment by mikeln** [comment]
Yeah, I have to second what Patrick said.  I think we need to adopt a "default is current" for anyplace with template subdirs, and then have version subdirs for past versions.
```
/templates
    whatever.jinja  <should be the v1.7 version>
    /v1.6
       whateverA.jinja 
    /v1.5
       whateverB.jinja
```
Also, from my initial look,  v1.7 should be created from v1.6 files.  (and for that matter, in almost all places it may be possible to make v1.6 files default, and remove the v1.6 directory.  In this way 1.6 and 1.7 will both use the same default file.  This would be perfectly ok, until some definite difference was found btwn 1.6 and 1.7).  

anything with 1.4 should be removed.  templates, variables, etc.

In kraken.config/files/config.yaml and gke-config.yaml.   We should be able to initially create a `<blah>17 `config at the `16` block. e.g.
```
 - &kubeVersionedFabric
      name: kubeVersionedFabric
      kind: versionedFabric
      type: canal
      kubeVersion:
        default: *defaultCanalFabric
        versions:
          v1.6: *defaultCanalFabric16
          v1.7: *defaultCanalFabric16
```
instead of creating a whole new block.  The only changes would be a version change, and those should apply to 16 also anyway.  The only time you would need a new 17 block is if the setup/vars are much different.

File: ansible/roles/kraken.access/tasks/generate-admin-kubeconfig.yaml (hunk 11 above)
Comment (joejulian) [inline ansible/roles/kraken.access/tasks/generate-admin-kubeconfig.yaml]: Perhaps this is the same as #667

Comment (mikeln) [inline ansible/roles/kraken.access/tasks/generate-admin-kubeconfig.yaml]: No.  I just threw together a 1.7 setup (completely based on 1.6).   It came up fine.  So these change lines are not needed.

**Comment by mikeln** [comment]
Actually, baring any changes to the directory structure to make `/template` defaults be 1.7, just duplicating the v1.6 directory to a new v1.7 directory is a good first step.   

You should not have to update DNS internal vars to _17. I think just running the 1.6 setup will work fine for now.

**Comment by coffeepac** [comment]
CI is failing because k2-tools doesn't have a 1.7 compatible version of helm installed.  I recommend updating build-scripts/update-generated-config.sh to set the config flag to use the latest version of helm that is installed.  Can probably tack it onto the `name` setting regex by doing a newline, spaces and setting  "helmOverride: true"  or something.  

I may have made this same comment on the k2-tools PR.

**Comment by samba** [comment]
@mikeln @coffeepac 
Based on comments here & discussion in Slack, I've outlined the process in `Documentation/UPGRADING_KUBERNETES.md`, for:

- Migrating (past) default templates to a version-specific template directory.
- Updating main cluster configuration to suit. 

I've also demonstrated it with the `v1.5` configuration templates.

Please review the outline and let me know if it sounds sensible. 

I'll follow up on the CI related tweaks to let this pass (e.g. helm version issues).

**Comment by samba** [comment]
Also, for what it's worth:

1. Previous attempt failed because I was basically forking `1.5` configuration and trying to hack it into a viable `1.7` cluster, without the benefit of all the `1.6` work done previously. (my bad)
2. With @mikeln's rendition, I've got functional `1.7.1` and `1.5.7` clusters running currently.

**Comment by samba** [comment]
(TO DO for @samba: verify that defaults match 1.7)
